### ISSUE TITLE
Task Field Forms UI

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#504]
Should be self explanatory.  This is the same logic for Technicians and Locations.  I have shown the Technicians example here.

"Available Technicians" in the screen shot below would be analogous to "Project Forms" from Issue #437.
Then Project Forms could be promoted to Task Forms (as Available Technicians could be promoted to Task Technicians.  see url:  /task/technicians).

![screen shot 2015-03-11 at 1 32 24 pm](https://cloud.githubusercontent.com/assets/8228624/6606331/0bbb7b98-c7f4-11e4-8cd7-13a9a979a4f6.png)

**Comment by damirius** [comment]
@bsaiken

I've added categorized lists with promotion buttons to new "Forms" tab in tasks. You can check it out like always on feature_damir branch. You'll also need to run 2015-03-17-1.sql db script to create the task_form relation table.
Since tasks are related to master_form and user_form records if you delete user_form from Field Forms UI list, relation between task and user_form will be deleted also and session will be updated. Just want to note here that since we use the same table for relationship between tasks and both user forms and master forms we must delete the records manually.
One small thing that I am curious about is "demoting" form from the Project on the form/listAll page. It would be in some way logical to delete the relationship between the form and the task if we remove the form from Project forms on that page. On the other hand maybe we want to leave the form linked to the task but remove the option to link other tasks to that form.
So to sum it up, if we demote the form on the Project Form UI do we need to remove that form from all already linked tasks?

**Comment by bsaiken** [comment]
@damirius 
This is a very good question - do we delete the task_form relationship when we delete the task_project relationship.   The answer is yes, we do need a cascade delete in this case.  However, it is not intuitively obvious.  The Project forms are templates - they never get filled out.  Their only purpose is to create child forms that will get completed and stored.  The purpose of creating the project_form and task_form relationships is to allow the user to set up Projects and Tasks.  Once a task has been completed, the user could, if they desired, delete the Project (and/or) Task, but the Completed Task, along with all the completed forms, would remain.

**Comment by damirius** [comment]
@bsaiken

I don't know if it's a typo there in your response, but the question was - do we delete the task_form relationship when we delete **project_form** relationship, not **task_project** relationship. Since we do actually have task_project relation but as 1 to many, task can't be linked to multiple projects just to one and we can't switch tasks between projects so only way to remove task from the project is to actually delete the task which will in turn delete all other relations automatically.
As your other part of the comment goes, since as far as I am aware, all of current forms are just templates we can discuss that part about uploading completed forms when we implement it.

**Comment by bsaiken** [comment]
@damirius 
Yes, deleted the task_form relationship if the project_form relationship is deleted.

**Comment by bsaiken** [comment]
@damirius 
The front end functionality is correct.  I leave it to you to assign the code review to @FWAJL in case you want to change the relation-deletion method (which is not really observed directly on the front end).

**Comment by damirius** [comment]
@bsaiken

You can actually recreate this through the front end.
Select some project as the current project.
Go to Field Forms -> List All to list all Project Forms.
Promote some Forms (either master or user) from Available Forms to Forms for this Project.
Go to Task List and choose some task to view, go to Forms tab.
You'll see all the template Forms in the list of Available Forms for that task. Promote one of them and remember the name of it.
Go back to Field Forms -> List All and demote the template form which you just promoted for the task.
Go back to that task and that Form will still be displayed in the "Forms for this task" left side panel.
If you demote it it will "disappear". Which is correct because it shouldn't be shown in the available forms since you demoted it from the list of project forms, but it is still shown in the left side panel because task_form relation is still in place.

In any case I'll see that I finish that up now.

Thanks!

**Comment by bsaiken** [comment]
@damirius 
Yes, I realize I can test it on the front end.  The reality is I am just trying to get on the road as today is a travel day.  
Technically the cascade deletion is not part of the scope for this Issue since I did not specify, but I'm glad you caught it, otherwise we'd be debugging later. :-)

**Comment by damirius** [comment]
@bsaiken

Not a problem. I've pushed new commit which takes care of that. I'll assign it to @FWAJL and I'll set it to code review.

Thanks

**Comment by FWAJL** [comment]
@bsaiken @damirius 

Ready for release.
